core/languages.lua

```diff
@@ -9,30 +13,124 @@ end)
 
 local loadonce = {}
 
+-- BEGIN OMIKHLEIA HACKLANG
+local icu = require("justenoughicu")
+
+-- This small utility could be moved to utilities as SU.forLanguage()...
+-- The input is expected to be a valid BCP47 canonical language.
+-- The idea is to find the "closest" language accepted by the callback:
+-- Loop removing a language specifier until the callback returns a non-nil value
+-- Returns that value and the matched language in that case, or nil no callback matched.
+-- E.g. "xx-Xxxx-XX" will be matched against "xx-Xxxx--XX", "xx-Xxxx", "xx" until one of
+-- these are satisfied.
+-- Returns
+--    nil if not callback could process the language
+--    language, res if a callback could (returning the matched language pattern and the
+--    result of the callback)
+local function forLanguage(langbcp47, callback)
+  while langbcp47 do
+    local res = callback(langbcp47)
+    if res then
+      return res, langbcp47
+    end
+    langbcp47 = langbcp47:match("^(.+)-.*$") -- split at dash (-) and remove last part.
+  end
+  return nil
+end
+-- END OMIKHLEIA HACKLANG
+
 SILE.languageSupport = {
-  languages = {},
+  languages = {}, -- BEGIN OMIKHLEIA HACKLANG
+    -- BAD CODE SMELL WARNING!!!!
+    -- In earlier versions this was used where we now have a "loadonce" table of boolean.
+    -- The change occurred here https://github.com/sile-typesetter/sile/commit/0c5e7f97f3c73ab1b6cd7aee0afca4a59c447cd9
+    -- So this table is not handled as it was, and shouldn't be used! But:
+    --   - font.lua uses it at one point...
+    --   - so does languages/kn.lua
+    --   - and also packages/complex-spaces
+    -- END OMIKHLEIA HACKLANG
   loadLanguage = function (language)
     language = language or SILE.settings:get("document.language")
-    language = cldr.locales[language] and language or "und"
+    -- BEGIN OMIKHLEIA HACKLANG
+    -- Either done too soon or plain wrong (a BCP47 language can e.g. contain a script for instance, sr-Latn
+    -- and I don't see that in https://github.com/alerque/cldr-lua/blob/master/cldr/data/locales.lua
+    -- I don't really know why we would want that CLDR filtering here BTW...
```

Comment:
From memory (obviously I'll be testing this here) one of the reasons for this is because using `icu.canonicalize_language()` gets us the canonical language code, but it does not validate that what we got could even be recognized, and just passes through garbage if you git it garbage. If my recollection is correct here, we should be doing both. First trying to parse a canonical name, then checking if that is a known language in the database and hence we can reasonably be expected to know anything about it.